src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,145 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.algebra.smooth_functions
+import ring_theory.derivation
+
+/-!
+
+# Derivation bundle
+
+In this file we define the derivations at a point of a manifold on the algebra of smooth fuctions.
+Moreover we define the total bundle of derivations (although at the moment it has not been given a
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+{M : Type*} [topological_space M] [charted_space H M]
+
+open_locale manifold
+
+namespace point_derivation
+
+instance smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := by apply_instance
+instance smooth_functions_tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := by apply_instance
+
+variables (𝕜 M) (n : with_top ℕ)
+
+/-- Type synonim to avoid metavariables. -/
+@[reducible, nolint unused_arguments] def pointed_smooth_map (x : M) := C^n⟮I, M; 𝕜⟯
+
+localized "notation `C^` n`⟮` I`,` M`;` 𝕜`⟯[` x`]` :=
+  point_derivation.pointed_smooth_map 𝕜 I M n x" in derivation
+
+variables {𝕜 M}
+
+/-- Evaluation at a point is a ring homomorphism. Same thing as writing manually
+`to_fun := λ f, f x`.-/
+def smooth_function.eval' (x : M) : C^∞⟮I, M; 𝕜⟯[x] →+* 𝕜 :=
+(pi.eval_ring_hom _ x : (M → 𝕜) →+* 𝕜).comp smooth_map.coe_fn_ring_hom
+
+variable {I}
+
+/-- The above evaluation gives rise to an algebra structure of `C^∞⟮I, M; 𝕜⟯` on `𝕜`. -/
+instance {x : M} : algebra C^∞⟮I, M; 𝕜⟯[x] 𝕜 := (smooth_function.eval' I x).to_algebra
+
+/-- With the above algebra structure evaluation is actually an algebra morphism. -/
+def smooth_function.eval (x : M) : C^∞⟮I, M; 𝕜⟯ →ₐ[C^∞⟮I, M; 𝕜⟯[x]] 𝕜 :=
```

Comment:
Isn't this usually an algebra morphism over `k` anyway, without any of this new structure? Composing smooth_map.coe_fn_alg_hom with pi.alg_hom_apply should construct that.

Reply:
I am not sure what you mean, `smooth_map.coe_fn_alg_hom` is a morphism of k algebras but what we want here is a morphism of `C^∞⟮I, M; 𝕜⟯[x]` algebras and `C^∞⟮I, M; 𝕜⟯[x]` is a strictly bigger ring than k, so the structure of algebra over `C^∞⟮I, M; 𝕜⟯[x]` cannot be recovered from that of algebra over `C^∞⟮I, M; 𝕜⟯[x]`. What did I miss?

Reply:
You can convert any `A →ₐ[𝕜] B` into a `A →ₐ [C^∞⟮I, M; 𝕜⟯[x]] B` via `alg_hom.restrict_scalars`, and I think all the typeclasses needed to do so are true.

My comment was largely confusion about your use of the word "actually" which to me implied the false "would not otherwise be"

Reply:
I am still confused: we have an inclusion morphism `𝕜 → C^∞⟮I, M; 𝕜⟯[x]` I can convert `A →ₐ [C^∞⟮I, M; 𝕜⟯[x]] B` into `A →ₐ [C^∞⟮I, M; 𝕜⟯[x]] B` and not the opposite. Am I getting confused? I might be too tired

Reply:
I think you have a typo in that message, as the expressions either side of `into` are identical.

At any rate the inclusion morphism is in the other direction, `eval' : C^∞⟮I, M; 𝕜⟯[x] →+* 𝕜` which you define above.

Reply:
Sorry I edited it

Reply:
I am really sorry to bother you with all these questions, but I still do not understand what you mean. To me an inclusion morphism is a monomorphism with an image, and `eval' : C^∞⟮I, M; 𝕜⟯[x] →+* 𝕜` is not an inclusion morphism as it is not a monomorphism. The inclusion morphism I was talking about is the constant functions morphism `𝕜 → C^∞⟮I, M; 𝕜⟯[x]`, which allows restricting scalars. If I use `eval' : C^∞⟮I, M; 𝕜⟯[x] →+* 𝕜` as my restrict scalars morphism I need an instance of `is_scalar_tower C^⊤⟮I,M;𝕜⟯[x] 𝕜 C^⊤⟮I, M; 𝓘(𝕜, 𝕜), 𝕜⟯` and this is simply mathematically false.

I think that the easiest way to explain to me what you want to do is to write down the line that you want me to write, so to spare us many other messages like these where I don't understand what you mean (I mean that line of code is for sure much shorter than all these messages we are writing to each other).

Reply:
> I need an instance of is_scalar_tower C^⊤⟮I,M;𝕜⟯[x] 𝕜 C^⊤⟮I, M; 𝓘(𝕜, 𝕜), 𝕜⟯ and this is simply mathematically false.

What definition would you like `has_scalar C^⊤⟮I,M;𝕜⟯[x] C^⊤⟮I, M; 𝕜⟯` to have? I think there are two choices.

Reply:
If they are the two that I am thinking of, they are not equal and the only definition that would make things work is the evaluation followed by multiplication! But I don't really see the point here, the proof is already one line and this instance at the moment would have no other purpose than simply changing the wording of the proof (which actually would become longer). I would wait to define such an instance when it actually has a real mathematical use in the theory.